adapter/postgres/postgres.go

```diff
@@ -47,103 +48,241 @@ func (d *PostgresDatabase) TableNames() ([]string, error) {
+		fmt.Fprintf(&queryBuilder, "%s;\n", primaryKeyDef)
+	}
+	for _, v := range indexDefs {
+		fmt.Fprintf(&queryBuilder, "%s;\n", v)
+	}
+	for _, v := range foreginDefs {
+		fmt.Fprintf(&queryBuilder, "%s;\n", v)
+	}
+	return strings.TrimSuffix(queryBuilder.String(), ";\n")
+}
 
-	// Ignore ALTER TABLE xxx OWNER TO yyy statements
-	re = regexp.MustCompilePOSIX("^ALTER TABLE [^ ;]+ OWNER TO .+;$")
-	ddl = re.ReplaceAllLiteralString(ddl, "")
+type column struct {
+	Name            string
+	dataType        string
+	Length          int
+	Nullable        bool
+	Default         string
+	IsPrimaryKey    bool
+	IsAutoIncrement bool
+	IsUnique        bool
+}
 
-	// Ignore REVOKE statements
-	re = regexp.MustCompilePOSIX("^REVOKE .*;$")
-	ddl = re.ReplaceAllLiteralString(ddl, "")
+func (c *column) GetDataType() string {
+	if c.IsAutoIncrement {
+		switch c.dataType {
+		case "smallint":
+			return "smallserial"
+		case "integer":
+			return "serial"
+		case "bigint":
+			return "bigserial"
+		}
+	}
+	return c.dataType
+}
 
-	// Ignore GRANT statements
-	re = regexp.MustCompilePOSIX("^GRANT .*;$")
-	ddl = re.ReplaceAllLiteralString(ddl, "")
+func (d *PostgresDatabase) getColumns(table string) ([]column, error) {
+	query := `SELECT column_name, column_default, is_nullable, data_type, character_maximum_length,
+	CASE WHEN p.contype = 'p' THEN true ELSE false END AS primarykey,
+	CASE WHEN p.contype = 'u' THEN true ELSE false END AS uniquekey
+FROM pg_attribute f
+	JOIN pg_class c ON c.oid = f.attrelid JOIN pg_type t ON t.oid = f.atttypid
+	LEFT JOIN pg_attrdef d ON d.adrelid = c.oid AND d.adnum = f.attnum
+	LEFT JOIN pg_namespace n ON n.oid = c.relnamespace
+	LEFT JOIN pg_constraint p ON p.conrelid = c.oid AND f.attnum = ANY (p.conkey)
+	LEFT JOIN pg_class AS g ON p.confrelid = g.oid
+	LEFT JOIN INFORMATION_SCHEMA.COLUMNS s ON s.column_name=f.attname AND c.relname=s.table_name
+WHERE c.relkind = 'r'::char AND c.relname = $1 AND f.attnum > 0 ORDER BY f.attnum;`
 
-	// Remove empty lines
-	// TODO: there should be a better way....
-	for strings.Replace(ddl, "\n\n", "\n", -1) != ddl {
-		ddl = strings.Replace(ddl, "\n\n", "\n", -1)
+	rows, err := d.db.Query(query, table)
+	if err != nil {
+		fmt.Println(err)
 	}
+	defer rows.Close()
 
-	ddl = strings.TrimSpace(ddl)
-	ddl = strings.TrimSuffix(ddl, ";") // XXX: caller should handle this better
+	cols := make([]column, 0)
+	for rows.Next() {
+		col := column{}
+		var colName, isNullable, dataType string
+		var maxLenStr, colDefault *string
+		var isPK, isUnique bool
+		err = rows.Scan(&colName, &colDefault, &isNullable, &dataType, &maxLenStr, &isPK, &isUnique)
+		if err != nil {
+			return nil, err
+		}
+		var maxLen int
+		if maxLenStr != nil {
+			maxLen, err = strconv.Atoi(*maxLenStr)
+			if err != nil {
+				return nil, err
+			}
+		}
+		col.Name = strings.Trim(colName, `" `)
+		if colDefault != nil || isPK {
+			if isPK {
+				col.IsPrimaryKey = true
+			} else {
+				col.Default = *colDefault
+			}
+		}
```

Comment:
Just curious, did you have any reason why this wasn't:

```go
		if isPK {
			col.IsPrimaryKey = true
		}
		if colDefault != nil {
			col.Default = *colDefault
		}
```
More specifically, did you intend to skip setting `col.Default` when `colDefault != nil` and `isPk` are true?